File: hub/codebuddy/bridge.js
#!/usr/bin/env node
'use strict';
/**
 * AgentProc bridge for the `codebuddy` CLI (Tencent CodeBuddy).
 *
 * CodeBuddy's stream-json output schema is compatible with claude's.
 * Differences: command name `codebuddy`, resume flag `-r`, env prefix CODEBUDDY_*.
 *
 * Mid-turn AgentProc permission is NOT supported: CodeBuddy documents
 * `--permission-prompt-tool` as unsupported. If AGENT_PERMISSION=1, exit with
 * AGENT_ERROR rather than silently falling back to skip-permissions.
 */

const path = require('node:path');
const HUB_DIR = path.resolve(__dirname, '..');
const {
  runBridge,
  emitError,
} = require(path.join(HUB_DIR, '_shared', 'stream_utils.js'));

const CLI_NAME = 'codebuddy';
const INSTALL_HINT = 'See your internal CodeBuddy installation docs.';
const PERMISSION_UNSUPPORTED =
  'codebuddy does not support mid-turn AgentProc permission ' +
  '(--permission-prompt-tool is documented as unsupported). ' +
  'Remove permission: true from the profile, or use hub/claude-code.';

function permissionEnabled(env) {
  return (env.AGENT_PERMISSION || '').trim() === '1';
}

function buildArgs(message, sessionId, env) {
  const args = [
    CLI_NAME, '-p', message,
    '--output-format', 'stream-json',
    '--dangerously-skip-permissions',
  ];
  const disallow = (env.CODEBUDDY_DISALLOW_TOOLS || 'AskUserQuestion').trim();
  if (disallow) {
    args.push('--disallowedTools', disallow);
  }
  const model = (env.CODEBUDDY_MODEL || '').trim();
  if (model) {
    args.push('--model', model);
  }
  if (sessionId) {
    args.push('-r', sessionId);
  }
  return args;
}

function parseEvent(event) {
  const etype = event.type;
  if (etype === 'assistant') {
    const text = (event.message?.content || [])
      .filter(b => b.type === 'text')
      .map(b => b.text)
      .join('');
    return text ? { partialText: text } : null;
  }
  if (etype === 'result') {
    const sessionId = event.session_id;
    if (event.is_error) {
      return { sessionId, error: event.result || 'codebuddy reported an error' };
    }
    const resultText = event.result || '';
    return { sessionId, finalText: resultText || null };
  }
  return null;
}

module.exports = { permissionEnabled, buildArgs, parseEvent, PERMISSION_UNSUPPORTED };

async function main() {
  if (permissionEnabled(process.env)) {
    emitError(PERMISSION_UNSUPPORTED);
    process.exit(1);
  }
  await runBridge({
    cliName: CLI_NAME,
    cliInstallHint: INSTALL_HINT,
    buildArgs,
    parseEvent,
  });
}

if (require.main === module) {
  main().catch(e => {
    process.stderr.write(`[codebuddy bridge] unhandled error: ${e && (e.stack || e)}\n`);
    process.exit(1);
  });
}

File: hub/codebuddy/bridge.py
#!/usr/bin/env python3
"""
AgentProc bridge for the `codebuddy` CLI (Tencent CodeBuddy, wire 0.3).

CodeBuddy's stream-json output schema is compatible with claude's. Differences:
    - command name: codebuddy
    - resume flag: -r <sessionId>  (instead of --resume)
    - env var prefix: CODEBUDDY_*  (instead of CLAUDE_*)

Mid-turn AgentProc permission is NOT supported: CodeBuddy documents
``--permission-prompt-tool`` as unsupported. If the turn carries
``permission: true``, the bridge emits an error rather than silently falling
back to skip-permissions.
"""

from __future__ import annotations

import os
import sys
from typing import Optional

_HUB_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
if _HUB_DIR not in sys.path:
    sys.path.insert(0, _HUB_DIR)

from _shared.stream_utils import EventResult, emit_error, main_entry, read_turn  # noqa: E402


CLI_NAME = "codebuddy"
INSTALL_HINT = "See your internal CodeBuddy installation docs."
PERMISSION_UNSUPPORTED = (
    "codebuddy does not support mid-turn AgentProc permission "
    "(--permission-prompt-tool is documented as unsupported). "
    "Remove permission: true from the profile, or use hub/claude-code."
)


def build_args(message: str, session_id: str, env) -> list[str]:
    args = [
        CLI_NAME, "-p", message,
        "--output-format", "stream-json",
        "--dangerously-skip-permissions",
    ]
    disallow = env.get("CODEBUDDY_DISALLOW_TOOLS", "AskUserQuestion")
    if disallow.strip():
        args += ["--disallowedTools", disallow]
    model = env.get("CODEBUDDY_MODEL", "").strip()
    if model:
        args += ["--model", model]
    if session_id:
        args += ["-r", session_id]
    return args


def parse_event(event: dict) -> Optional[EventResult]:
    etype = event.get("type")
    if etype == "assistant":
        text = "".join(
            b.get("text", "")
            for b in (event.get("message") or {}).get("content", [])
            if b.get("type") == "text"
        )
        return EventResult(partial_text=text) if text else None
    if etype == "result":
        session_id = event.get("session_id")
        if event.get("is_error"):
            return EventResult(
                session_id=session_id,
                error=event.get("result", "codebuddy reported an error"),
            )
        result_text = event.get("result", "")
        return EventResult(
            session_id=session_id,
            final_text=result_text if result_text else None,
        )
    return None


def main() -> int:
    turn = read_turn()
    if turn.get("permission") is True:
        emit_error(PERMISSION_UNSUPPORTED)
        return 1
    return main_entry(CLI_NAME, INSTALL_HINT, build_args, parse_event, turn=turn)


if __name__ == "__main__":
    sys.exit(main())
